src/png/mod.rs

```diff
@@ -112,27 +119,46 @@ impl PngData {
                 b"IHDR" | b"PLTE" | b"tRNS" => {
                     key_chunks.insert(chunk.name, chunk.data.to_owned());
                 }
-                _ => {
-                    if opts.strip.keep(&chunk.name) {
-                        if chunk.is_c2pa() {
-                            // StripChunks::None is the default value, so to keep optimizing by default,
-                            // interpret it as stripping the C2PA metadata.
-                            // The C2PA metadata is invalidated if the file changes, so it shouldn't be kept.
-                            if opts.strip == StripChunks::None {
-                                continue;
-                            }
-                            return Err(PngError::C2PAMetadataPreventsChanges);
+                _ if opts.strip.keep(&chunk.name) => {
+                    if chunk.is_c2pa() {
+                        // StripChunks::None is the default value, so to keep optimizing by default,
+                        // interpret it as stripping the C2PA metadata.
+                        // The C2PA metadata is invalidated if the file changes, so it shouldn't be kept.
+                        if opts.strip == StripChunks::None {
+                            continue;
+                        }
+                        return Err(PngError::C2PAMetadataPreventsChanges);
+                    }
+                    if chunk.name == *b"fcTL" || chunk.name == *b"fdAT" {
+                        // Validate the sequence number
+                        if read_be_u32(&chunk.data[0..4]) != sequence_number {
+                            return Err(PngError::APNGOutOfOrder);
+                        }
+                        sequence_number += 1;
+                        if chunk.name == *b"fcTL" && !idat_data.is_empty() {
+                            // Only create a Frame if it's after the IDAT (else store it as an aux chunk)
+                            frames.push(Frame::from_fctl_data(chunk.data)?);
+                            continue;
```

Comment:
The [PNG spec](https://www.w3.org/TR/png-3/#structure) allows `fcTL` to appear before `IDAT` to use the image encoded in `IDAT` as the first frame of the animation:

> The static image may be included as the first frame of the animation by the presence of a single [fcTL](https://www.w3.org/TR/png-3/#fcTL-chunk) chunk before [IDAT](https://www.w3.org/TR/png-3/#11IDAT). Otherwise, the static image is not part of the animation.

So, handling such a `fcTL` as a regular auxiliar, ancillary chunk may change the displayed animation if it gets stripped like other ancillary chunks, which does not seem intended since `fcTL` is indeed a chunk critical for display in the context of APNG.

What was your rationale to write this condition this way? :thinking: :slightly_smiling_face:

Reply:
So basically at the point we're at right now on the journey towards full APNG support, this just seemed like the easiest way of dealing with that. Note that this is after we've performed the keep/strip check so we know fcTLs are being kept here (stripping of all APNG chunks is permitted).

If we did construct a Frame from an initial fcTL then we'd need additional logic both to handle an "empty" frame and to make sure that it gets written in the correct place in the output.

Reply:
Hmm, alright, that makes sense. We can reasonably improve on this later on, so no worries for now :+1: